### PR TITLE
Use Strict in hlist.LeftFolder and hlist.RightFolder to avoid diverging implicit expansion

### PR DESCRIPTION
**Pull request by DmytroMitin**
Details are here: https://stackoverflow.com/questions/52175141/diverging-implicit-expansion-error-on-foldleft-hlist/

`RightFolder` works properly without `Lazy`.

### CODE REVIEW

**Comment by milessabin** [comment]
Do you have a test which fails before this PR and succeeds afterwards?

Why have you enabled `scala.language.reflectiveCalls`?

**Comment by DmytroMitin** [comment]
@milessabin 
Before PR the following code doesn't compile:
```
object poly extends Poly2 {
  implicit def wrapMap[T]: Case.Aux[T, Int, Map[List[Int], T]] = at[T, Int]((end, t) => Map(List(t) -> end)) 
}

implicitly[LeftFolder.Aux[Int :: Int :: HNil, Int, poly.type, Map[List[Int], Map[List[Int], Int]]]]
```

After PR it does.

Without `scala.language.reflectiveCalls` test `poly.scala` doesn't compile.

https://travis-ci.org/milessabin/shapeless/jobs/424845742

```
[error] /home/travis/build/milessabin/shapeless/core/src/test/scala/shapeless/poly.scala:427:56: reflective access of structural type member method allCases should be enabled
[error] by making the implicit value scala.language.reflectiveCalls visible.
[error] This can be achieved by adding the import clause 'import scala.language.reflectiveCalls'
[error] or by setting the compiler option -language:reflectiveCalls.
[error] See the Scaladoc for value scala.language.reflectiveCalls for a discussion
[error] why the feature should be explicitly enabled.
[error]     val r = (10 :: "hello" :: true :: HNil).foldLeft(0)(myPoly)
[error]                                                        ^
```

**Comment by joroKr21** [comment]
It works in Scala 2.13. `Strict` might be a better choice here.

**Comment by joroKr21** [comment]
Also `RightFolder` might have the same problem and this needs a test.

File: core/src/test/scala/shapeless/poly.scala [hunk 1]
```diff
@@ -16,6 +16,8 @@
 
 package shapeless
 
+import scala.language.reflectiveCalls
+
```
Comment (joroKr21) [inline core/src/test/scala/shapeless/poly.scala]:
```suggestion
```

File: core/src/main/scala/shapeless/ops/hlists.scala [hunk 2]
```diff
@@ -429,10 +429,10 @@ object hlist {
       }
 
     implicit def hlistLeftFolder[H, T <: HList, In, HF, OutH, FtOut]
-      (implicit f : Case2.Aux[HF, In, H, OutH], ft : LeftFolder.Aux[T, OutH, HF, FtOut]): Aux[H :: T, In, HF, FtOut] =
+      (implicit f : Case2.Aux[HF, In, H, OutH], ft : Lazy[LeftFolder.Aux[T, OutH, HF, FtOut]]): Aux[H :: T, In, HF, FtOut] =
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/ops/hlists.scala]:
```suggestion
      (implicit f : Case2.Aux[HF, In, H, OutH], ft : Strict[LeftFolder.Aux[T, OutH, HF, FtOut]]): Aux[H :: T, In, HF, FtOut] =
```

**Comment by milessabin** [comment]
Apparently there really is some unexpected reflective access there ...

**Comment by joroKr21** [comment]
> Apparently there really is some unexpected reflective access there ...

I have no clue why this is happening

**Comment by milessabin** [comment]
Missing explicit refinement somewhere maybe?

**Comment by joroKr21** [comment]
@milessabin I figured out the cause of the reflective calls warning - `PolyNBuilders` were missing type annotations. Check the last commit.

**Comment by milessabin** [comment]
> I figured out the cause of the reflective calls warning

Good catch!

**Pull request merged**